3.11 builds failing with v2.8.1 on manylinux2010 images

### Description

With:
```yaml
        cibw_python: [ "cp37-*", "cp38-*", "cp39-*", "cp310-*", "cp311-*"]
```
and
```yaml
        env:
          CIBW_PRERELEASE_PYTHONS: True
```

I'm seeing this error:
```
Setting up build environment...
  
      + /opt/python/cp38-cp38/bin/python -c 'import sys, json, os; json.dump(os.environ.copy(), sys.stdout)'
      + which python
  cibuildwheel: python available on PATH doesn't match our installed instance. If you have modified PATH, ensure that you don't overwrite cibuildwheel's entry or insert python above it.
  Error: Process completed with exit code 1.
```

This didn't occur with 2.8.0 - I suspect it's somehow related to this changelog entry:

> 🛠 The GitHub Action will ensure a compatible version of Python is installed on the runner (#1114)

I'm also seeing
```
Run actions/setup-python@v4
  with:
    python-version: 3.7 - 3.10
    update-environment: false
    token: ***
  env:
    CIBW_BUILD: cp311-*
    CIBW_ARCHS_LINUX: x86_64
    CIBW_MANYLINUX_X86_64_IMAGE: manylinux[20](https://github.com/bloomberg/memray/runs/7420878603?check_suite_focus=true#step:3:21)10
    CIBW_MANYLINUX_I686_IMAGE: manylinux2010
    CIBW_PRERELEASE_PYTHONS: true
    CIBW_TEST_EXTRAS: test
    CIBW_TEST_COMMAND: pytest {package}/tests
```
in the build log - perhaps that "python-version: 3.7 - 3.10" is the problem?

### Build log

https://github.com/bloomberg/memray/runs/7421778954?check_suite_focus=true

### CI config

https://github.com/bloomberg/memray/blob/f9bfaacea95c9c565189da2eeb121a3303c11c71/.github/workflows/build_wheels.yml

--- a/cibuildwheel/linux.py
+++ b/cibuildwheel/linux.py
@@ -110,6 +110,29 @@ def get_build_steps(
     yield from steps.values()
 
 
+def check_all_python_exist(
+    *, platform_configs: list[PythonConfiguration], container: OCIContainer
+) -> None:
+    exist = True
+    messages = []
+    for config in platform_configs:
+        python_path = config.path / "bin" / "python"
+        try:
+            container.call(["test", "-x", python_path])
+        except subprocess.CalledProcessError:
+            messages.append(
+                f"  '{python_path}' executable doesn't exist in image '{container.image}' to build '{config.identifier}'."
+            )
+            exist = False
+    if not exist:
+        message = "\n".join(messages)
+        print(
+            f"cibuildwheel:\n{message}",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+
 def build_in_container(
     *,
     options: Options,
@@ -119,6 +142,8 @@ def build_in_container(
     container_package_dir: PurePath,
 ) -> None:
     container_output_dir = PurePosixPath("/output")
+
+    check_all_python_exist(platform_configs=platform_configs, container=container)
 
     log.step("Copying project into container...")
     container.copy_into(Path.cwd(), container_project_path)

--- a/test/test_linux_python.py
+++ b/test/test_linux_python.py
@@ -1,0 +1,43 @@
+from __future__ import annotations
+
+import platform
+import subprocess
+
+import pytest
+
+from . import test_projects, utils
+
+
+def test_python_exist(tmp_path, capfd):
+    if utils.platform != "linux":
+        pytest.skip("the test is only relevant to the linux build")
+    machine = platform.machine()
+    if machine not in ["x86_64", "i686"]:
+        pytest.skip(
+            "this test is currently only possible on x86_64/i686 due to availability of alternative images"
+        )
+
+    project_dir = tmp_path / "project"
+    basic_project = test_projects.new_c_project()
+    basic_project.generate(project_dir)
+
+    with pytest.raises(subprocess.CalledProcessError):
+        utils.cibuildwheel_run(
+            project_dir,
+            add_env={
+                "CIBW_MANYLINUX_X86_64_IMAGE": "manylinux2010",
+                "CIBW_MANYLINUX_I686_IMAGE": "manylinux2010",
+                "CIBW_BUILD": "cp3{10,11}-manylinux*",
+            },
+        )
+
+    captured = capfd.readouterr()
+    print("out", captured.out)
+    print("err", captured.err)
+    assert f" to build 'cp310-manylinux_{machine}'." not in captured.err
+    message = (
+        "'/opt/python/cp311-cp311/bin/python' executable doesn't exist"
+        f" in image 'quay.io/pypa/manylinux2010_{machine}:2022-08-05-4535177'"
+        f" to build 'cp311-manylinux_{machine}'."
+    )
+    assert message in captured.err

